core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java

```diff
@@ -0,0 +1,288 @@
+    mAsyncBlockRemover = new AsyncBlockRemover(mBlockWorker);
+    mBlockMasterSyncHelper = new BlockMasterSyncHelper(mMasterClient);
+    mMasterAddress = masterClient.getRemoteSockAddress();
+    mBlockHeartbeatReporter = heartbeatReporter;
+  }
+
+  private void registerWithMaster() {
+    RetryPolicy retry = createEndlessRetry();
+    while (retry.attempt()) {
+      try {
+        LOG.info("Registering with master {}", mMasterAddress);
+        // The content in the report can be cleared because registration will
+        // report these block information anyways.
+        mBlockHeartbeatReporter.clear();
+        registerWithMasterInternal();
+        LOG.info("Finished registration with {}", mMasterAddress);
+        return;
+      } catch (Exception e) {
+        LOG.error("Failed to register with master {}, error {}, retry count {} Will retry...",
+            mMasterAddress, e, retry.getAttemptCount());
+        mWorkerState = WorkerMasterRegistrationState.NOT_REGISTERED;
+      }
+    }
+    // Should not reach here because the retry is indefinite
+    ProcessUtils.fatalError(LOG, new RuntimeException(),
+        "Failed to register with master %s", mMasterAddress);
+  }
+
+  protected void registerWithMasterInternal()
+      throws IOException, FailedToAcquireRegisterLeaseException {
+    // The target master is not necessarily the one that allocated the workerID
+    LOG.info("Notify the master {} about the workerID {}", mMasterAddress, mWorkerId);
+    mMasterClient.notifyWorkerId(mWorkerId.get(), mWorkerAddress);
+    // TODO(elega) If worker registration to all masters happens at the same time,
+    // this might cause OOM issue. Consider limiting the worker registration concurrency.
+    BlockStoreMeta storeMeta = mBlockWorker.getStoreMetaFull();
+
+    try {
+      mBlockMasterSyncHelper.tryAcquireLease(mWorkerId.get(), storeMeta);
+    } catch (FailedToAcquireRegisterLeaseException e) {
+      if (Configuration.getBoolean(PropertyKey.TEST_MODE)) {
+        throw new RuntimeException(String.format("Master register lease timeout exceeded: %dms",
+            ACQUIRE_LEASE_WAIT_MAX_DURATION));
+      }
+      throw e;
+    }
+    mWorkerState = WorkerMasterRegistrationState.REGISTERING;
+    mBlockMasterSyncHelper.registerToMaster(mWorkerId.get(), storeMeta);
+
+    mWorkerState = WorkerMasterRegistrationState.REGISTERED;
+    Metrics.WORKER_MASTER_REGISTRATION_SUCCESS_COUNT.inc();
+    mLastSuccessfulHeartbeatMs = CommonUtils.getCurrentMs();
+  }
+
+  private RetryPolicy createEndlessRetry() {
+    return new ExponentialBackoffRetry(
+        1000, 60 * 1000, Integer.MAX_VALUE);
+  }
+
+  @Override
+  public synchronized void heartbeat() throws InterruptedException {
+    if (mWorkerState == WorkerMasterRegistrationState.NOT_REGISTERED) {
+      // Not registered because:
+      // 1. The worker just started, we kick off the 1st registration here.
+      // 2. Master sends a registration command during
```

Comment:
3. the heartbeat message becomes too big that we decide to fall back to a full re-register